[FIX] core: add support for windows-874 encoding

Armin-FalDiS opened the pull request:
Description of the issue/feature this PR addresses:
Outlook and similar Windows based systems use windows-874 for encoding Thai characters which is not natively known by Python. Simply aliasing the Windows encoding as cp874 adds support for this encoding.

Current behavior before PR:
If we send an email from outlook with Thai characters, this error is thrown when Odoo receives the mail:
`LookupError: unknown encoding: windows-874`

Desired behavior after PR is merged:
Thai characters which are encoded with `windows-874` encoding are now supported in Odoo either by email or other means.


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/158060).

Julien00859 commented on addons/test_mail/tests/test_mail_gateway.py:
```diff
@@ -18,6 +18,7 @@
 from odoo.tests.common import users
 from odoo.tools import email_split_and_format, formataddr, mute_logger
 
+from email.header import Header
```
I'm sorry to be this guy but can you move this import next to the `unittest` ones?

```py
...
import socket

from email.header import Header
from unittest.mock import DEFAULT
from unittest.mock import patch
...

Julien00859 replied:
and since you're at it, you can change the commit title to `[FIX] core: add support for windows-874 encoding`

Armin-FalDiS replied:
WDM if it's better/cleaner it should be moved. Thanks!

Julien00859 commented on addons/test_mail/tests/test_mail_gateway.py:
```diff
@@ -1703,18 +1704,18 @@ def test_message_hebrew_iso8859_8_i(self):
         # is iso-8859-8 which is natively supported by python.
-        charset = "iso-8859-8"
-        content = "שלום וברוכים הבאים למקרה המבחן הנפלא הזה"
+        content = 'שלום וברוכים הבאים למקרה המבחן הנפלא הזה'
```
useless diff (for the content line, moving the charset one is fine)

Julien00859 commented on addons/test_mail/tests/test_mail_gateway.py:
```diff
@@ -1703,18 +1704,18 @@ def test_message_hebrew_iso8859_8_i(self):
         with RecordCapturer(self.env['mail.test.gateway'], []) as capture:
             mail = test_mail_data.MAIL_FILE_ENCODING.format(
-                msg_id="<test_message_hebrew_iso8859_8_i@iron.sky>",
+                msg_id='<test_message_hebrew_iso8859_8_i@iron.sky>',
```
useless diff

Julien00859 reviewed (changes requested):
In stable we try not to change lines unless it is necessary. Changing that `encoded_subject` to use the `Header` class is kind of already a violation of that rule but the risk/benefit is very low. Let's not PEP8 the whole file (also we could argue about what quote single vs double we shall use in this context, let's just not)

Don't forget that:

> and since you're at it, you can change the commit title to [FIX] core: add support for windows-874 encoding

Julien00859 reviewed (changes requested):
Third time. `[FIX]`, not `[IMP]`. We are fixing a bug (ok, a python one, but a bug nonetheless) not adding a new feature.

Armin-FalDiS commented:
> Third time. `[FIX]`, not `[IMP]`. We are fixing a bug (ok, a python one, but a bug nonetheless) not adding a new feature.

Oh apologies. I thought I copied the entire line from your comment.

Julien00859 commented on odoo/__init__.py:
```diff
@@ -85,6 +85,14 @@ def _decompress(data):
+
+encodings.aliases.aliases['874'] = 'cp874'
+encodings.aliases.aliases['windows_874'] = 'cp874'
+encodings.aliases.aliases['windows-874'] = 'cp874'
```
No need to add an alias for `windows-874`. You can register onle `windows_874` and `windows-874` will work out of the box. https://github.com/python/cpython/blob/15309329b65a285cb7b3071f0f08ac964b61411b/Lib/encodings/__init__.py#L43-L69

Julien00859 commented on addons/test_mail/tests/test_mail_gateway.py:
```diff
@@ -1728,6 +1729,32 @@ def test_message_hebrew_iso8859_8_i(self):
+        subject = 'ขออนุญาตทบทวนเอกสารโครงการ'
+        encoded_subject = Header(subject, charset)
+        content = 'แมวสีดำเป็นสัตว์เลี้ยงที่น่ารักมากๆ แ=ละมีเสน่ห์ในตัวอย่างมาก ๆ ด้วย'
+        encoded_content = base64.b64encode(content.encode(charset)).decode()
```
with some more time thinking about it I no longer think that it is a good idea to use `Header` and `content.encode(charset)`. We should provide with the actual bytes found in the customer's email (ask for their EML) and not engineer our way to reproduce such strings.

Here the problem is that we added that `windows-874` alias and we are using it to both encode + decode the text. We could replace `charset = 'windows-874'` by `charset = 'utf-8'` the test would green.

In the original test (the hebrew one) we shouldn't had use `encoded_content = base64.b64encode(content.encode(charset)).decode()` in the first place but instead provide with actual bytes extracted out of a real-world email. This `encoded_content` is testing when the encoded text is found in the body of the email / inside an attachment, I think we had no real world example at hand and decided to craft one ourself.

Julien00859 replied:
So I suggest we put our hand on the EML file that doesn't work inside odoo (you can ask the help of the tech sm squad, they are used to this kind of situation) and base our test on that file.

Armin-FalDiS commented:
Hey @Julien00859 
I agree with the fact that in it's previous state, the test was not really testing anything!
So I have now added the EML file to Odoo and now everything should be in order. I have also removed the changes on the other test. I believe we should stick to the guidelines as you said and change that test in a separate commit.

Let me know what you think. Thanks!

Julien00859 reviewed (approved):
It now looks great, just revert that last change inside the hebrew test, I'll sort it myself in another PR don't worry :)

@robodoo delegate+

Armin-FalDiS commented:
Hey @Julien00859 
Done. I also change the emails on the EML data. I don't think having my actual email there is a good idea xD.
